File: src/Server/AppTheme.cs
namespace Belin.UI;

using System.Text.Json.Serialization;

/// <summary>
/// Enumerates different themes an operating system or application can show.
/// </summary>
[JsonConverter(typeof(JsonStringEnumConverter))]
public enum AppTheme {

	/// <summary>
	/// The system theme.
	/// </summary>
	System,

	/// <summary>
	/// The light theme.
	/// </summary>
	Light,

	/// <summary>
	/// The dark theme.
	/// </summary>
	Dark
}

/// <summary>
/// Provides extension members for application themes.
/// </summary>
public static class AppThemeExtensions {
	extension(AppTheme appTheme) {

		/// <summary>
		/// The icon corresponding to this theme.
		/// </summary>
		public string Icon => appTheme switch {
			AppTheme.System => "contrast",
			AppTheme.Light => "light_mode",
			AppTheme.Dark => "dark_mode"
		};

		/// <summary>
		/// The label corresponding to this theme.
		/// </summary>
		public string Label => appTheme switch {
			AppTheme.System => "Auto",
			AppTheme.Light => "Clair",
			AppTheme.Dark => "Sombre"
		};
	}
}

File: src/Client/AppTheme.ts
/**
 * Enumerates different themes an operating system or application can show.
 */
export const AppTheme = Object.freeze({

	/**
	 * The system theme.
	 */
	System: "System",

	/**
	 * The light theme.
	 */
	Light: "Light",

	/**
	 * The dark theme.
	 */
	Dark: "Dark"
});

/**
 * Enumerates different themes an operating system or application can show.
 */
export type AppTheme = typeof AppTheme[keyof typeof AppTheme];

/**
 * Gets the icon corresponding to the specified theme.
 * @param theme The application theme.
 * @returns The icon corresponding to the specified theme.
 */
export function icon(theme: AppTheme): string {
	switch (theme) {
		case AppTheme.Dark: return "dark_mode";
		case AppTheme.Light: return "light_mode";
		default: return "contrast";
	}
}

/**
 * Gets the label corresponding to the specified theme.
 * @param theme The application theme.
 * @returns The label corresponding to the specified theme.
 */
export function label(theme: AppTheme): string {
	switch (theme) {
		case AppTheme.Dark: return "Sombre";
		case AppTheme.Light: return "Clair";
		default: return "Auto";
	}
}
